README.md
# ReaModuleSecurity Anypoint Connector

[Connector description including destination service or application with]

# Mule supported versions
Examples:
Mule 3.4.x, 3.5.x
Mule 3.4.1

# [Destination service or application name] supported versions
Example:
Oracle E-Business Suite 12.1 and above.

#Service or application supported modules
Example:
Oracle CRM
Oracle Financials
or 
Salesforce API v.24
Salesforce Metadata API


# Installation 
For beta connectors you can download the source code and build it with devkit to find it available on your local repository. Then you can add it to Studio

For released connectors you can download them from the update site in Anypoint Studio. 
Open Anypoint Studio, go to Help → Install New Software and select Anypoint Connectors Update Site where you’ll find all avaliable connectors.

#Usage
For information about usage our documentation at http://github.com/mulesoft/rea-module-security.

# Reporting Issues

We use GitHub:Issues for tracking issues with this connector. You can report new issues at this link http://github.com/mulesoft/rea-module-security/issues.

CHANGELOG.md
# ReaModuleSecurity Release Notes
# --------------------------------------------------------------------
# Date: <DD-MMM-YYYY>
Example: 24-Jan-2015
# Version: (default value: 0.1)
# Supported API versions: 
Example: Salesforce Bulk API v.28, SOAP API V.28 (include links to API docs)
# Supported Mule Runtime Versions: 
${project.devkitVersion}
# New Features and Functionality
(default value: Initial version) 
Example: Added support for Query. http://github.com/mulesoft/rea-module-security/issues/61
# Closed Issues in this release 
(include issue link when possible)
Example: Fixed problem with lost telefone on Bulk Update (CLDCONNECT-123).
# Known Issues in this release
Example: Session ID expires after  30 days. http://github.com/mulesoft/rea-module-security/issues/62


src/test/java/au/com/reagroup/security/test/TestJAXRS.java
package au.com.reagroup.security.test;

import javax.ws.rs.client.Client;
import javax.ws.rs.client.ClientBuilder;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import org.glassfish.jersey.client.authentication.HttpAuthenticationFeature;

public class TestJAXRS {

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		
		//HttpAuthenticationFeature feature = HttpAuthenticationFeature.basic("username", "password");
		 
		final Client client = ClientBuilder.newClient();
		//client.register(feature);

		Response response = client.target("http://localhost:8081/api").
				path("user").path("authenticate").request(MediaType.APPLICATION_JSON).header("Authorization", "blah").get();
		
		System.out.println(response.getStatus());
	    System.out.println(response.getStatusInfo());
	    
	     
	    if(response.getStatus() == 200)
	    {
	    	System.out.println(response.getHeaderString("X-Authenticated-userid"));
	        
	    }

	    //UserResource.GetUserAuthenticateResponse.accepted().
    	
		

	}

}


src/test/java/au/com/reagroup/security/test/automation/runner/FunctionalTestSuite.java
package au.com.reagroup.security.test.automation.runner;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.runner.RunWith;
import org.junit.runners.Suite;
import org.junit.runners.Suite.SuiteClasses;
import au.com.reagroup.security.test.automation.functional.AuthenticateUserTestCases;
import au.com.reagroup.security.test.automation.functional.AuthenticateUserWithGroupTestCases;
import au.com.reagroup.security.SecurityConnector;
import org.mule.tools.devkit.ctf.mockup.ConnectorTestContext;

@RunWith(Suite.class)
@SuiteClasses({ AuthenticateUserTestCases.class, AuthenticateUserWithGroupTestCases.class })

public class FunctionalTestSuite {

	@BeforeClass
	public static void initialiseSuite() {
		ConnectorTestContext.initialize(SecurityConnector.class);
	}

	@AfterClass
	public static void shutdownSuite() {
		ConnectorTestContext.shutDown();
	}

}

src/test/java/au/com/reagroup/security/test/automation/functional/AuthenticateUserTestCases.java
package au.com.reagroup.security.test.automation.functional;

import static org.junit.Assert.*;
import au.com.reagroup.security.SecurityConnector;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mule.tools.devkit.ctf.junit.AbstractTestCase;

public class AuthenticateUserTestCases extends AbstractTestCase<SecurityConnector> {

	public AuthenticateUserTestCases() {
		super(SecurityConnector.class);
	}

	@Before
	public void setup() {
		// TODO
	}

	@After
	public void tearDown() {
		// TODO
	}

	@Test
	public void verify() throws Exception {
		java.lang.Object expected = null;
		org.mule.api.callback.SourceCallback callback = null;
		org.mule.api.MuleEvent event = null;
		org.mule.api.MuleMessage message = null;
		java.lang.String auth = null;
		assertEquals(getConnector().authenticateUser(callback, event, message, auth), expected);
	}

}

src/main/java/au/com/reagroup/security/SecurityConnector.java
package au.com.reagroup.security;

import java.net.Inet4Address;
import java.net.UnknownHostException;
import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.mule.api.MuleContext;
import org.mule.api.MuleEvent;
import org.mule.api.MuleMessage;
import org.mule.api.annotations.Category;
import org.mule.api.annotations.Config;
import org.mule.api.annotations.Connector;
import org.mule.api.annotations.Processor;
import org.mule.api.annotations.display.FriendlyName;
import org.mule.api.annotations.display.Placement;
import org.mule.api.annotations.param.Default;
import org.mule.api.callback.SourceCallback;
import org.mule.api.transport.PropertyScope;

import au.com.reagroup.security.config.ADSecurityConfig;



@Connector(name="rea-module-security", friendlyName="REA Group Security Module")
@Category(name = "org.mule.tooling.category.security", description = "Security")
public class SecurityConnector extends AbstractBaseModule {
	
	private static Logger LOGGER = LogManager.getLogger(SecurityConnector.class);
	
	private static String HOSTNAME;
	
    static {
        if (System.getenv("COMPUTERNAME") != null) {
            HOSTNAME = System.getenv("COMPUTERNAME");
        } else if (System.getenv("HOSTNAME") != null) {
            HOSTNAME = System.getenv("HOSTNAME");
        } else {
            try {
                HOSTNAME = Inet4Address.getLocalHost().getHostName();
            } catch (UnknownHostException e) {
                HOSTNAME = "[unknown]";
            }
        }
    }
	
    @Config
    ADSecurityConfig config;

    public ADSecurityConfig getConfig() {
		return config;
	}

	public void setConfig(ADSecurityConfig config) {
		this.config = config;
	}

	/**
     * The Mule context is injected to access metadata such as the flow name.
     */
	@Inject
    private MuleContext muleContext;

    public void setMuleContext(MuleContext muleContext) {
        this.muleContext = muleContext;
    }

	/**
     * Validate User
     *
     * @param 
     * @return The MuleEvent
     */
    @Processor(intercepting = true)
    @Category(name = "Validate User", description = "Validate username/password")
    public Object authenticateUser(
    		SourceCallback callback, 
            MuleEvent event,
            MuleMessage message,

            @FriendlyName("Basic Authentication (Base64 Format)")
            @Placement(order = 1, group = "Settings", tab = "General")
            @Default("#[message.inboundProperties['Authorization']]")
            String auth) throws Exception {
    	
    	LOGGER.debug(String.format(SecurityUtil.LOG_MESSAGE_DEBUG_FORMAT, message.getUniqueId(), "Entering authenticateUser"));
    	
    	Map<String, Object> mapa = new HashMap<String, Object>();

    	String flowName = event.getFlowConstruct().getName();
    	
    	LOGGER.info(HOSTNAME);
    	LOGGER.info("Hit Authenticate User for Flow Name: "+flowName);
    	LOGGER.info("Uri="+getConfig().getServiceEndpoint());
    	
    	try {
    	
    		String userId = authenticateUser(auth, getConfig().getServiceEndpoint());
    		mapa.put(SecurityUtil.REA_AUTHENTICATED_USER_HEADER, userId);
    		event.getMessage().addProperties(mapa, PropertyScope.OUTBOUND);
    		callback.process(event.getMessage().getPayload());
    		return event.getMessage().getPayload();
    		
    	} catch (UnauthorizedException e) {
    		mapa.put("WWW-Authenticate", "Basic realm=\"Unauthorized\"");
			mapa.put("http.status", 401);
			event.getMessage().addProperties(mapa, PropertyScope.OUTBOUND);
			event.getMessage().setPayload(String.format(SecurityUtil.JSON_RESPONSE_FORMAT, e.getMessage()));
			return event.getMessage().getPayload();
    	
    	} catch (Exception e) {
			mapa.put("http.status", 500);
			event.getMessage().addProperties(mapa, PropertyScope.OUTBOUND);
			event.getMessage().setPayload(String.format(SecurityUtil.JSON_RESPONSE_FORMAT, e.getMessage()));
			return event.getMessage().getPayload();
    	} finally {
    		LOGGER.debug(String.format(SecurityUtil.LOG_MESSAGE_DEBUG_FORMAT, message.getUniqueId(), "Leaving authenticateUser"));
    	}
    	 	

    }
    
	/**
     * Validate User with Group
     *
     * @param 
     * @return The MuleEvent
     */
    @Processor(intercepting = true)
    @Category(name = "Validate User with Group", description = "Validate username/password and also whether the user is a member of the specified group")
    public Object authenticateUserWithGroup(
    		SourceCallback callback, 
            MuleEvent event,
            MuleMessage message,
            
            @FriendlyName("Basic Authentication (Base64 Format)")
            @Placement(order = 1, group = "Settings", tab = "General")
            @Default("#[message.inboundProperties['Authorization']]")
            String auth,

            @FriendlyName("Group Name")
            @Placement(order = 1, group = "Settings", tab = "General")
            String group) {

    	LOGGER.debug(String.format(SecurityUtil.LOG_MESSAGE_DEBUG_FORMAT, message.getUniqueId(), "Entering authenticateUser"));
    	
    	Map<String, Object> mapa = new HashMap<String, Object>();

    	String flowName = event.getFlowConstruct().getName();
    	
    	LOGGER.info(HOSTNAME);
    	LOGGER.info("Hit Authenticate User for Flow Name: "+flowName);
    	LOGGER.info("Uri="+getConfig().getServiceEndpoint());
    	
    	try {
    	
    		String userId = authenticateUserWithGroup(auth, group, getConfig().getServiceEndpoint());
    		mapa.put(SecurityUtil.REA_AUTHENTICATED_USER_HEADER, userId);
    		event.getMessage().addProperties(mapa, PropertyScope.OUTBOUND);
    		callback.process(event.getMessage().getPayload());
    		return event.getMessage().getPayload();
    		
    	} catch (UnauthorizedException e) {
    		mapa.put("WWW-Authenticate", "Basic realm=\"Unauthorized\"");
			mapa.put("http.status", 401);
			event.getMessage().addProperties(mapa, PropertyScope.OUTBOUND);
			event.getMessage().setPayload(String.format(SecurityUtil.JSON_RESPONSE_FORMAT, e.getMessage()));
			return event.getMessage().getPayload();
    	} catch (ForbiddenException e) {
			mapa.put("http.status", 403);
			event.getMessage().addProperties(mapa, PropertyScope.OUTBOUND);
			event.getMessage().setPayload(String.format(SecurityUtil.JSON_RESPONSE_FORMAT, e.getMessage()));
			return event.getMessage().getPayload();
    	} catch (Exception e) {
			mapa.put("http.status", 500);
			event.getMessage().addProperties(mapa, PropertyScope.OUTBOUND);
			event.getMessage().setPayload(String.format(SecurityUtil.JSON_RESPONSE_FORMAT, e.getMessage()));
			return event.getMessage().getPayload();
    	} finally {
    		LOGGER.debug(String.format(SecurityUtil.LOG_MESSAGE_DEBUG_FORMAT, message.getUniqueId(), "Leaving authenticateUser"));
    	}
    }
    
    


}

src/main/java/au/com/reagroup/security/config/ADSecurityConfig.java
package au.com.reagroup.security.config;

import javax.ws.rs.client.Client;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.mule.api.ConnectionException;
import org.mule.api.annotations.Configurable;
import org.mule.api.annotations.Connect;
import org.mule.api.annotations.ConnectionIdentifier;
import org.mule.api.annotations.Disconnect;
import org.mule.api.annotations.TestConnectivity;
import org.mule.api.annotations.ValidateConnection;
import org.mule.api.annotations.components.ConnectionManagement;
import org.mule.api.annotations.display.FriendlyName;
import org.mule.api.annotations.display.Placement;
import org.mule.api.annotations.param.ConnectionKey;
import org.mule.api.annotations.param.Default;

import au.com.reagroup.security.SecurityUtil;

@ConnectionManagement(configElementName="rea-security-config", friendlyName = "Configuration")
public class ADSecurityConfig {
	
	private static Logger LOGGER = LogManager.getLogger(ADSecurityConfig.class);
	
    /**
     * REA Group Security App URI
     */
    @Configurable
    @Placement(group = "Connection")
    @Default("https://rea-auth.rea-group.com:8092/api")
    private String serviceEndpoint;

	public String getServiceEndpoint() {
		return serviceEndpoint;
	}

	public void setServiceEndpoint(String reaSecurityServiceUri) {
		this.serviceEndpoint = reaSecurityServiceUri;
	}
	
	@Connect
	@TestConnectivity
	public void connect(@ConnectionKey @Default("https://rea-auth.rea-group.com:8092/api") @FriendlyName("AD API Connection Url") String testUrl) 
			throws ConnectionException {
		setServiceEndpoint(testUrl);
		
		LOGGER.debug("Testing REA Active Directory API Connection: "+serviceEndpoint);
		Client client = SecurityUtil.getClientConnection();
		Response response = client.target(serviceEndpoint).
				path("user").path("authenticate").request(MediaType.APPLICATION_JSON).get();
		LOGGER.info("Got: "+response.getStatus());
		
	}
	
	@Disconnect
	public void disconnect() {}
	
    @ConnectionIdentifier
	public String identifier() {
		
		return "";
	}
	
	@ValidateConnection
	public boolean connection() {
		
		return true;
	}

}


src/main/java/au/com/reagroup/security/AbstractBaseModule.java
package au.com.reagroup.security;

import java.net.ConnectException;

import javax.ws.rs.client.Client;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public abstract class AbstractBaseModule {
	
	private static Logger LOGGER = LogManager.getLogger(AbstractBaseModule.class);
	
	/**
	 * Function to Validate a User
	 * 
	 * @param target
	 * @return
	 */
	public String authenticateUser(String auth, String target) throws UnauthorizedException, ConnectException, Exception {
		
		Client client = null;
		Response response = null;
		
		try {
		
			client = SecurityUtil.getClientConnection();
	
			response = client.target(target).
					path("authenticate").request(MediaType.APPLICATION_JSON).
					header(SecurityUtil.AUTHORIZATION_HTTP_HEADER, auth).get();
			
		} catch (javax.ws.rs.ProcessingException e) {
			
			if (e.getCause() instanceof ConnectException) {
				throw new ConnectException(String.format("Unable to connect to REA Authentication API - %s",e.getMessage()));
			} else {
				throw e;
			}
		} 
	    
		/*
		 * Check HTTP Status Codes
		 */
	    if(response.getStatus() == 200)
	    {
	    	LOGGER.info("got 200");
	    	return response.getHeaderString(SecurityUtil.REA_AUTHENTICATED_USER_HEADER); 
	    } 
	    else if (response.getStatus() == 401)
	    {
	    	LOGGER.info("got 401");
	    	throw new UnauthorizedException("Not authorized to access resource");
	    } else {
	    	LOGGER.info("got "+response.getStatus());
	    	throw new RuntimeException(String.format("REA Authentication API returned HTTP Status Code: [%s]",response.getStatus()));
	    }
	    
	}
	
	/**
	 * Function to Validate a User against an LDAP group
	 * 
	 * @param auth
	 * @param group
	 * @param target
	 * @return
	 * @throws UnauthorizedException
	 * @throws ConnectException
	 * @throws Exception
	 */
	public String authenticateUserWithGroup(String auth, String group, String target) throws UnauthorizedException, ForbiddenException, ConnectException, Exception {
		Client client = null;
		Response response = null;
		
		try {
		
			client = SecurityUtil.getClientConnection();
	
			response = client.target(target).
					path("authenticate").path(group).request(MediaType.APPLICATION_JSON).
					header(SecurityUtil.AUTHORIZATION_HTTP_HEADER, auth).get();
			
		} catch (javax.ws.rs.ProcessingException e) {
			
			if (e.getCause() instanceof ConnectException) {
				throw new ConnectException(String.format("Unable to connect to REA Authentication API - %s",e.getMessage()));
			} else {
				throw e;
			}
		} 
	    
		/*
		 * Check HTTP Status Codes
		 */
	    if(response.getStatus() == 200)
	    {
	    	LOGGER.info("got 200");
	    	return response.getHeaderString(SecurityUtil.REA_AUTHENTICATED_USER_HEADER); 
	    } 
	    else if (response.getStatus() == 401)
	    {
	    	LOGGER.info("got 401");
	    	throw new UnauthorizedException("Not authorized to access resource");
	    }
	    else if (response.getStatus() == 403) {
	    	LOGGER.info("got 403");
	    	throw new ForbiddenException("User is not a member of the group");
	    } else {
	    	LOGGER.info("got "+response.getStatus());
	    	throw new RuntimeException(String.format("REA Authentication API returned HTTP Status Code: [%s]",response.getStatus()));
	    }
	}

}
